tangelo/toolboxes/molecular_computation/fno.py

```diff
@@ -0,0 +1,422 @@
+from tangelo.toolboxes.operators import FermionOperator
+
+
+class FNO:
+    """Class to interface with the Frozen Natural Orbitals protocol, that aims
+    at reducing the computational cost of a molecular problem by truncating
+    the virtual space. In general, the virtual orbitals are ranked according to
+    their MP2 occupancies, and selected with a given threshold. They are also
+    transformed to the FNO basis, using the eigenvectors of the MP2
+    virtual-virtual density matrix.
+
+    Attributes:
+            sqmol (SecondQuantizedMolecule): Self-explanatory.
+            uhf (bool): Flag indicating the type of mean field used.
+            n_mos (int): Number of molecular orbitals.
+            fock_ao (np.array): Fock matrix in atomic orbital form.
+            frozen_occupied (list): List of indices of frozen occupied orbitals.
+            threshold (float or list): Threshold(s) for FNO occupancy.
+
+        Properties:
+            fermionic_hamiltonian: Returns a fermionic Hamiltonian, with the
+                frozen orbitals and updated MO coefficients.
+
+        Methods:
+            compute_fno: Computes and truncates the FNO orbitals.
+            get_frozen_indices: Determines the indices of the frozen orbitals.
+            diagonalize_and_reorder: Diagonalizes a matrix and can reorder the
+                eigenvalues and eigenvectors based on occupations.
+            get_number_of_fnos_from_frac_occupancies: Calculates the number of
+                active virtual orbitals based on fractional occupancies.
+    """
+
+    def __init__(self, sqmol: SecondQuantizedMolecule, threshold: Union[float, Sequence[float]]):
+        """Initialization of the FNO class instance.
+
+        Checks for frozen virtual orbitals and warns if they are set, as they
+        might be overwritten.
+
+        Args:
+            sqmol (SecondQuantizedMolecule): The SecondQuantizedMolecule
+                object.
+            threshold (float or list): Threshold(s) for FNO occupancy.
+        """
+
+        # Check if there are frozen virtual orbitals. Print warning that this
+        # setting will be overwritten.
+        if sqmol.frozen_virtual and sqmol.frozen_virtual != [[], []]:
+            warnings.warn(f"The frozen orbitals indices will be overwritten in {self.__class__.__name__}.", RuntimeWarning)
+
+        self.sqmol = sqmol
+
+        # Molecular properties useful for the FNO class methods.
+        self.uhf = sqmol.uhf
+        self.n_mos = self.sqmol.n_mos
+        self.fock_ao = self.sqmol.mean_field.get_fock()
+        self.frozen_occupied = self.sqmol.frozen_occupied
+
+        # Verify threshold format.
+        if isinstance(threshold, float) and 0. < threshold <= 1.:
+            self.threshold = [threshold] * 2 if self.uhf else threshold
+        elif isinstance(threshold, (list, tuple, np.ndarray)) and self.uhf and all(map(lambda x: 0. < x <= 1., threshold)):
```

Comment:
It's usually better to force the input to be a specific type, unless you believe that all these types may arise naturally as part of users workflow.

Too many options makes it confusing to the user and can lead to insidious bugs or make maintenance of code / compatibility difficult.

I let you exert your best judgment here and decide whether this is good as it is or not (for example you may know that indeed a tuple could get in there)